File: 2019/day1/day1.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func readInput(fname string) (input []int, err error) {
	bytes, err := ioutil.ReadFile(fname)
	if err != nil {
		return nil, err
	}

	lines := strings.Split(string(bytes), "\n")
	input = make([]int, 0, len(lines))
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}
		num, err := strconv.Atoi(line)
		if err != nil {
			return nil, err
		}
		input = append(input, num)
	}

	return input, nil
}

func calcFuel1(mass int) (fuel int) {
	return (mass / 3) - 2
}

func part1(input []int) (fuel int) {
	fuel = 0
	for i := 0; i < len(input); i++ {
		fuel += calcFuel1(input[i])
	}
	return fuel
}

func calcFuel2(mass int) (fuel int) {
	fuel = (mass / 3) - 2

	if fuel > 0 {
		fuel += calcFuel2(fuel)
		return fuel
	} else {
		return 0
	}
}

func part2(input []int) (fuel int) {
	fuel = 0
	for i := 0; i < len(input); i++ {
		fuel += calcFuel2(input[i])
	}
	return fuel
}

func main() {
	input, err := readInput("day1/input.txt")
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Println("part1: ", part1(input))
	fmt.Println("part2: ", part2(input))
}

File: 2019/day1/day1.mjs
import fs from 'fs';

function readInput(fname) {
    const content = fs.readFileSync(fname, 'utf8');
    return content
        .split('\n')
        .filter(l => l)
        .map(l => parseInt(l, 10));
}

function calcFuel1(mass) {
    return Math.floor(mass / 3) - 2;
}

function part1(input) {
    return input.reduce((fuel, mass) => fuel + calcFuel1(mass), 0);
}

function calcFuel2(mass) {
    const fuel = Math.floor(mass / 3) - 2;
    const additional = Math.max(fuel > 0 ? calcFuel2(fuel) : 0, 0);
    return Math.max(fuel + additional, 0);
}

function part2(input) {
    return input.reduce((fuel, mass) => fuel + calcFuel2(mass), 0);
}

function main() {
    const input = readInput('day1/input.txt');
    console.log('Part1: ', part1(input));
    console.log('Part2: ', part2(input));
}

main();
